src/hooks/usePermalink.ts

```diff
@@ -71,38 +73,194 @@ interface HookResult {
+ */
+const determineUserId = (
+    type: PillType | null,
+    parseResult: PermalinkParts | null,
+    event: MatrixEvent | null,
+): string | null => {
+    if (type === null) return null;
+
+    if (parseResult?.userId) return parseResult.userId;
+
+    if (event && [PillType.EventInSameRoom, PillType.EventInOtherRoom].includes(type)) {
+        return event.getSender() ?? null;
+    }
+
+    return null;
+};
+
+/**
+ * Tries to find a room by room Id or searching all rooms for an alias.
+ *
+ * @param roomIdOrAlias - Id or alias of the room to find.
+ * @returns Room if found, else null.
+ */
+const findRoom = (roomIdOrAlias: string): Room | null => {
+    const client = MatrixClientPeg.get();
+
+    return roomIdOrAlias[0] === "#"
+        ? client.getRooms().find((r) => {
+              return r.getCanonicalAlias() === roomIdOrAlias || r.getAltAliases().includes(roomIdOrAlias);
+          }) ?? null
+        : client.getRoom(roomIdOrAlias);
+};
+
+/**
+ * Tries to determine the initial room.
+ * Initial here means it should be possible to load the room without sending API requests.
+ * For an @room or a user mention it is the permalinkRoom.
+ * If the parse result contains a room Id or alias try to find it with {@link findRoom}.
+ * Otherwise returns null.
+ *
+ * @param type - Pill type
+ * @param permalinkRoom - Room in which the permalink is displayed.
+ * @param parseResult - Permalink parser result
+ * @returns Initial room or null if it cannot be determined.
+ */
+const determineInitialRoom = (
+    type: PillType | null,
+    permalinkRoom: Room | undefined,
+    parseResult: PermalinkParts | null,
+): Room | null => {
+    if (type === PillType.AtRoomMention && permalinkRoom) return permalinkRoom;
+
+    if (type === PillType.UserMention && permalinkRoom) {
+        return permalinkRoom;
+    }
+
+    if (parseResult?.roomIdOrAlias) {
+        const room = findRoom(parseResult.roomIdOrAlias);
+        if (room) return room;
+    }
+
+    return null;
+};
+
+/**
+ * Tries to find the initial event.
+ * If the event should not be looked up or there is no target room it returns null.
+ * Otherwise it tries to get the event from the target room.
+ *
+ * @param shouldLookUpEvent - whether the parmalink event should be looked up
+ * @param Room | null targetRoom - target room of the permalink
+ * @param parseResult - permalink parse result
+ * @returns The event if found or null if it should not be looked up or was not found.
+ */
+const determineInitialEvent = (
+    shouldLookUpEvent: boolean,
+    targetRoom: Room | null,
+    parseResult: PermalinkParts | null,
+): MatrixEvent | null => {
+    if (!shouldLookUpEvent || !targetRoom || !parseResult?.eventId) return null;
+
+    return targetRoom.findEventById(parseResult.eventId) || null;
+};
+
+/**
+ * Can be used to retrieve all information needed to display a permalink.
+ */
+export const usePermalink: (args: Args) => HookResult = ({
```

Comment:
This hook is long and I wonder if we can split the content in multiple custom hooks and use them inside this hook. It'll improve the readability of the hook